servers/zms/src/main/java/com/yahoo/athenz/zms/DBService.java

```diff
@@ -3537,4 +3548,137 @@ public void executePutRoleSystemMeta(ResourceContext ctx, String domainName, Str
+                    throw ex;
+                }
+            }
+        }
+    }
+
+    /**
+     * If the role has audit enabled, and user did not provide the auditRef,
+     * an exception will be thrown. This is the first check before any write
+     * operation is carried out so we don't really have anything to roll-back
+     **/
+    Role checkRoleAuditEnabled(ObjectStoreConnection con, String domainName, String roleName, String auditRef, String caller, String principal) {
+
+        Role role = con.getRole(domainName, roleName);
+        if (role == null) {
+            con.rollbackChanges();
+            throw ZMSUtils.notFoundError(caller + ": Unknown role: " + roleName + " in domain: " + domainName, caller);
+        }
+
+        if (role.getAuditEnabled()) {
+            if (auditRef == null || auditRef.length() == 0) {
+                con.rollbackChanges();
+                throw ZMSUtils.requestError(caller + ": Audit reference required for role: " + domainName, caller);
+            }
+
+            if (auditReferenceValidator != null && !auditReferenceValidator.validateReference(auditRef, principal, caller)) {
+                con.rollbackChanges();
+                throw ZMSUtils.requestError(caller + ": Audit reference validation failed for domain: " + domainName + " and role: " + roleName
+                        + ", auditRef: " + auditRef, caller);
+            }
+        }
+
+        return role;
+    }
+
+    void executePutMembershipDecision(ResourceContext ctx, String domainName, String roleName,
+                              RoleMember roleMember, String auditRef, String caller) {
+
+        // our exception handling code does the check for retry count
+        // and throws the exception it had received when the retry
+        // count reaches 0
+
+        for (int retryCount = defaultRetryCount; ; retryCount--) {
+
+            try (ObjectStoreConnection con = store.getConnection(true, true)) {
+
+                String principal = getPrincipalName(ctx);
+
+                // process our insert role member support. since this is a "single"
+                // operation, we are not using any transactions.
+
+                if (!con.confirmRoleMember(domainName, roleName, roleMember,
+                        principal, auditRef)) {
+                    con.rollbackChanges();
+                    throw ZMSUtils.requestError(caller + ": unable to apply role membership decision for member: " +
+                            roleMember.getMemberName() + " and role: " + roleName, caller);
+                }
+
+                // update our role and domain time-stamps, and invalidate local cache entry
+
+                con.updateRoleModTimestamp(domainName, roleName);
+                con.updateDomainModTimestamp(domainName);
+                cacheStore.invalidate(domainName);
+
+                // audit log the request
+
+                auditLogRequest(ctx, domainName, auditRef, caller, ZMSConsts.HTTP_PUT,
```

Comment:
this doesn't seem sufficient since it doesn't say if the request was approved or denied.